### PR TITLE
Z3 v4.8.7 update

### PR DESCRIPTION
**Pull request by baierd**
Discuss the update of Z3 to v4.8.7 as well as the issues #168  #154 #116  and #148 here.
Z3 does only provide binaries for Ubuntu 16.04 and Windows at this point, so we will have to compile it ourselfs.
The JavaSMT code for interpolation was already removed and the tests adapted.
@kfriedberger

### CODE REVIEW

**Comment by PhilippWendler** [comment]
Why can't we use the binary for Ubuntu 16.04? Is this too new for what JavaSMT wants to support?

**Comment by kfriedberger** [comment]
The binary for Ubuntu16.04 should also work on newer Ubuntu versions. It should be fine.
@baierd please test and update the scripts, afterwards I will update the Ivy-part and merge this PR.

**Comment by kfriedberger** [comment]
I fixed the ant script for building Z3 from its release and updated the Z3 version in Ivy up to 4.8.7.
There are several SegFaults in the JUnit tests (at least on my laptop).

@baierd 
Can you take a look? Is this a problem in Z3 or on our site?

**Comment by baierd** [comment]
I accidentally reverted my Z3 lib, hence why i did not catch that.

There is a change in the return value of the parser that needs fixing (It returns a vector of
formulas as opposed to a conjunction of formulas now).
I will fix this tomorrow.
Additionally i will add solver interrupt and smt2 logging (because the Native module is able to use these after all)

**Comment by kfriedberger** [comment]
Currently, we need to fix the model evaluation for Z3 (high priority).
Additionally, we can extend the current functionality in JavaSMT with new features like SMT-logging in Z3 (low priority).

**Comment by baierd** [comment]
I fixed the parser and i am working on the model at the moment.

File: src/org/sosy_lab/java_smt/solvers/z3/Z3SolverContext.java [hunk 1]
```diff
@@ -162,6 +162,7 @@ public static synchronized Z3SolverContext create(
       Native.setParamValue(cfg, "PROOF", "true");
     }
     Native.globalParamSet("smt.random_seed", String.valueOf(randomSeed));
+    Native.globalParamSet("model.compact", "false");
```
Comment (kfriedberger) [inline src/org/sosy_lab/java_smt/solvers/z3/Z3SolverContext.java]: @baierd can you give more details on this option? is this option disabling a `partial model`?

Comment (baierd) [reply src/org/sosy_lab/java_smt/solvers/z3/Z3SolverContext.java]: It deactivates a new default option that compresses function definitions. Deactivating restores the old standard behavior.

**Comment by kfriedberger** [comment]
The logfile is currently empty for my own tests.

**Comment by kfriedberger** [comment]
@baierd Could you update the documentation for Z3?
At least the following places seem to be outdated:
- https://github.com/sosy-lab/java-smt/blob/master/doc/Developers.md#publishing-z3
- https://github.com/sosy-lab/java-smt/blob/master/doc/KnownIssues.md#z3

**Comment by baierd** [comment]
Sure, ill take a look.

**Comment by baierd** [comment]
@kfriedberger 
The SMT2-Log had a bug that is now fixed in Z3.
Should i remove the SMT2-Log until the next release of Z3?

**Comment by kfriedberger** [comment]
Yes, remove the log from JavaSMT as long as it does not work properly.
If possible, add a reference to the bug report in Z3 into this discussion.
Then we can merge this PR or is anything remaining?

**Comment by kfriedberger** [comment]
@baierd any news?

**Comment by baierd** [comment]
The Z3 Log is removed.
I think we can merge.

Z3 Issue for the SMT2 Log: [https://github.com/Z3Prover/z3/issues/3513](url)

**Pull request merged**